### PR TITLE
cache + speedup windows e2e

### PR DESCRIPTION
**Pull request by penalosa**
## What this PR solves / how to test

This adds additional turbo caching to e2e builds, as well as speeding up Wrangler installs when running e2e for Windows

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [ ] Included
  - [x] Not necessary because: testing change
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): <https://github.com/cloudflare/cloudflare-docs/pull/>...
  - [x] Not necessary because: internal change

### CODE REVIEW

**Review by petebacondarwin — APPROVED** [review]

File: .github/workflows/changesets.yml
Comment (petebacondarwin) [inline .github/workflows/changesets.yml:42]: NIT: the description of this step is not very helpful. Can we describe the intent instead?

```suggestion
      - name: "Configure turbo repo cache environment variables"
```

File: .github/workflows/e2e.yml
Comment (petebacondarwin) [inline .github/workflows/e2e.yml:72]: No need for abbreviation here

```suggestion
      - name: Find pnpm global directory
```

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
There are quite a few changes to what trigger different jobs to run. Please can you update `.github/workflows/README.md` with this information.
Can you also explain the reasoning behind moving some tests from PRs to main? We recently did the opposite. One of the reasons was that it was not as easy to be notified if there is a failure on main compared to on a PR (such as Version Packages) and such failed jobs don't block PRs from landing. Perhaps I am misunderstanding the changes here.

File: vitest.shared.ts
Comment (CarmenPopoviciu) [inline vitest.shared.ts:22]: I'm not against this change, but was wondering if adding these retries won't increase our build times even more

**Comment by penalosa** [comment]
> Can you also explain the reasoning behind moving some tests from PRs to main?

The reason for moving these to main is that on the version packages PR the versions of packages to be released are bumped. This invalidates the turbo cache (and isn't configurable), and so we'd lose the benefit of caching for running tests before release. With our new process of locking main pre release, we'll just have to make sure that main is green before merging the changesets PR.

File: vitest.shared.ts
Comment (penalosa) [reply vitest.shared.ts:22]: This only affects packages which run tests with vitest, which are all so fast it doesn't make much of a difference (it excludes wrangler and miniflare, for instance). I think the extra robustness should make up for potentially longer build times

File: .github/workflows/test-and-check.yml
Comment (CarmenPopoviciu) [inline .github/workflows/test-and-check.yml:7]: is this left "empty" on purpose?

**Comment by CarmenPopoviciu** [comment]
with the relatively limited context I have, this LGTM. Left all the comments I had for things that stood out. Let's pls get one more approval before we land this, and thank you for working on it Samuel <3

**Comment by petebacondarwin** [comment]
Discussed offline: we agreed not to run the tests on main but leave the on the Version Packages PR.
Needs a bit of a rebase.

**Review by petebacondarwin — APPROVED** [review]
Magic!

**Pull request merged**
